Refactor CalendarDatePicker and Update Fluent Style

robloo opened the pull request:
## What does the pull request do?

**Note: As this is a bigger change than I wanted, and is tearing into Microsoft code, anyone on the core team please let me know if this type of refactoring is going to be an issue sooner rather than later.**

CalendarDatePicker's style is not considered production-ready as noted in #7928. It looks like the style was hacked on-top of existing Microsoft code that was used as a base for the Calendar. 

This code is some of the worst-quality I have seen in the project and is very difficult to maintain/update. This caused a problem when trying to update the style and it was realized it wasn't using pseudoclasses at all and had a lot of hardcoded/missing functionality. So a hack on top of a hack at this point.

Therefore, I am begrudgingly fixing things correctly.

## Open Questions

 1. The calendar button itself can be removed. This is now redundant functionality that isn't needed. In the UWP implementation it does not exist -- however, I currently keep it for backwards compatiblity.
 1. Keeping the calendar as a button may cause issues with navigation especially by keyboard.
 1. There is an alternate way of constructing this control. Perhaps the entire control itself could just be a button. Then the TextBox is part of the content just like the icon. This might solve the above two issues. The style would have to be heavily customized though.

## What is the current behavior?

Poorly designed CalendarDatePicker Fluent style sitting on top of poorly integrated Microsoft calendar code.

## What is the updated/expected behavior with this PR?

 1. Updated the CalendardDatePicker style to be production ready closing #7928
     * Fixed poor icon contrast when selected/pointer over (can't see black text over dark blue background). Text colors now properly change in all states and the blue backgrounds are gone.
     * Calendar icon is now inside the input box area like other XAML frameworks
     * The date is now vertically centered in the input TextBox
     * Disabled state is now fully implemented
 3. In attempting to do number 1, it was discovered that psudoclasses would need to be added and used (the control was never properly integrated with Avalonia in the past). Added :pressed and :flyout-open
 4. In attempting to do number 2, a moderate refactoring was needed as the code was poor quality and hard to modify adding the needed functionality.
 5. Added missing Fluent v1 resources for light/dark theme
 6. Support opening the calendar flyout by pressing on the CalendarDatePicker itself (previously the button -- and only the button -- must be pressed).
 7. While looking at source code I fixed `Button` which was missing some calls to the base methods in some overrides
 8. Also modernized `TextBox` to use latest convention for property accessors

| Before | After |
|--|--|
| ![image](https://user-images.githubusercontent.com/17993847/162789488-3942d618-5a4c-4b78-a9a6-fa5355058bb0.png) | ![image](https://user-images.githubusercontent.com/17993847/162639231-4c353864-ac0b-4d27-9ceb-38f881efef40.png) |

## How was the solution implemented (if it's not obvious)?

As described.

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes

The Fluent style of the CalendardDatePicker is no longer compatible with past versions. Now the template parts didn't change so this shouldn't technically break anything; however, custom styles will need to be updated.

## Obsoletions / Deprecations

None

## Fixed issues

Closes #7928

robloo committed:
Update property definitions to latest conventions

robloo committed:
Update XAML formatting

robloo committed:
Move CalendarDatePicker properties into partial class

robloo committed:
Use OnPropertyChanged overridable method

robloo committed:
Remove single-use methods and process within OnPropertyChanged

robloo committed:
Remove unused methods

robloo committed:
Move CalendarDatePickerDateValidationErrorEventArgs into its own file

robloo committed:
Move CalendarDatePickerFormat into its own file

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0019784-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Add :pressed and :flyout-open PseudoClasses to CalendarDatePicker

timunie commented:
Hi @robloo I hope you are okay with asking a question. While you update this control, do you think we can implement calendar weeks?

robloo commented:
@timunie No plans to do so and I have to watch my time commitments. That said, I have a future plan for Calendar itself [#calendar](https://github.com/AvaloniaUI/Avalonia/issues/5656) that will make it as fast and easy to use as a DatePicker. In fact it could have replaced DatePicker entirely if Microsoft designed things differently from the start. Calendar weeks are something that could probably be done at the same time.

timunie commented:
Well I have this done by attached behavior for me, but thought this would be a great opportunity to port it over for everyone. I would share my code for sure if you were interested or had the time. But I fully understand your point of view. 👍🙂. I even don't know if that is something that a calendar should have, as also WPF does not offer this. 

Have a great evening

robloo commented:
@timunie Feel free to open an issue for it. I may not understand entirely what you are talking about if it can be done with an attached property only. I thought you wanted a week-view of the calendar itself (I know certain cultures/industries like to refer to week numbers directly). If you only want the week number property `int` 1-52 that is fairly simple (as opposed to a new view). It should probably be separate from this PR though.

robloo committed:
Rewrite the Fluent CalendarDatePicker style

robloo committed:
Add watermark property comments

robloo committed:
Fix references in comments

robloo committed:
Modernize TextBox property accessor code

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0019785-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Make sure to call base methods in Button

robloo committed:
Support opening the calendar flyout when pressing outside the button

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0019787-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Better support MinHeight

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0019788-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Implement focused states in CalendarDatePicker

robloo commented:
@amwx This redesign of the template and styles should make Fluent v2 styling in FluentAvalonia easier. However, if you need anything added for FluentAvalonia please let me know.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0019789-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Improve CalendarDatePicker Fluent style

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0019793-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Merge branch 'master' into calendar-date-picker

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0019918-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Merge branch 'master' into calendar-date-picker

robloo committed:
Use more standard x:Double resource

robloo committed:
Remove Microsoft copyright

An audit of git blame shows the style has now been fully rewritten for Avalonia

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0019989-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Merge branch 'master' into calendar-date-picker

robloo committed:
Update code for generic virtual method changes

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020148-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Merge branch 'master' into calendar-date-picker

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020154-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo commented:
@maxkatz6 This has been ready for review for a while. There are open questions to discuss but they may not drive any changes. 

If this is going to be open a while longer, I'm going to add one more fix to remove the MaxWidth. It will be around 1 week before I get to having time for Avalonia commits though.

maxkatz6 commented:
@robloo I will review this PR this Thursday. Thanks for reminding me.

maxkatz6 committed:
Merge branch 'master' into calendar-date-picker

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020371-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

maxkatz6 commented on src/Avalonia.Controls/Calendar/CalendarDatePicker.cs:
```diff
@@ -538,16 +197,200 @@ protected override void OnApplyTemplate(TemplateAppliedEventArgs e)
+            // Text
+            else if (change.Property == TextProperty)
+            {
+                var oldValue = change.GetOldValue<string>();
+                var value = change.GetNewValue<string>();
```
minor:
Now you can use ```change.GetOldAndNewValue<string>();```

maxkatz6 commented on src/Avalonia.Controls/Calendar/CalendarDatePicker.cs:
```diff
@@ -538,16 +197,200 @@ protected override void OnApplyTemplate(TemplateAppliedEventArgs e)
+            // Text
+            else if (change.Property == TextProperty)
+            {
+                var oldValue = change.GetOldValue<string>();
+                var value = change.GetNewValue<string>();
```
minor: Now you can use
```suggestion
                var (oldValue, value) = change.GetOldAndNewValue<string>();
```

robloo replied:
Thanks, I never looked up all the methods and extensions that were available in this type of code (actually I think there were too many options).

maxkatz6 replied:
It's a new method in master branch

maxkatz6 commented on src/Avalonia.Themes.Fluent/Controls/CalendarDatePicker.xaml:
```diff
@@ -18,131 +11,172 @@
+                      Grid.Column="1"
+                      Focusable="False" />
+              <Popup Name="PART_Popup"
+                     PlacementTarget="{TemplateBinding}"
```
Unexpected, didn't know this can work. But seems it does!

robloo replied:
That code wasn't from me -- I'm very surprised it works too. I think I'll add the path just to be clear.

robloo replied:
I didn't read this properly the first time. Obviously, it's similar to the `{Binding}` syntax which is perfectly valid. `{Binding}` just means bind to the current data context. `{TemplateBinding}` is similar and just means bind to the TemplateBinding context which is the parent (`TemplateBinding` is equivalent to `{Binding RelativeSource={RelativeSource TemplatedParent}`). This seems like a decent way of doing things and should just be documented somewhere. So I'm not going to change anything although I could probably replace it with the relative source equivalent binding.

Note that the Popup automatically finds a good target if one isn't set. So setting this isn't even required.

Note: I'm going into this level of detail more to document this for myself later.

maxkatz6 commented on src/Avalonia.Themes.Fluent/Controls/CalendarDatePicker.xaml:
```diff
@@ -18,131 +11,172 @@
+                       VerticalContentAlignment="{TemplateBinding VerticalContentAlignment}"
+                       HorizontalContentAlignment="{TemplateBinding HorizontalContentAlignment}"
+                       Grid.Column="0" />
+              <Button Name="PART_Button"
+                      Grid.Column="1"
+                      Focusable="False" />
```
This button is not focusable, and I don't see how user can open flyout.
That's probably old behavior, but I would expect "Alt+ArrowDown" to work here, as it works in SplitButton. Might be done separately.

robloo replied:
I can add that functionality without too much trouble I think. Will be next week before I circle back to commits though.

robloo replied:
Done to open the Popup/Flyout only. Escape is required to close it. This is because the Calendar handles the down key and it just creates a bad user experience. Further details in comments.

maxkatz6 commented:
> This code is some of the worst-quality I have seen in the project and is very difficult to maintain/update

Yeah, but NumericUpDown is definitely worse.
Anyway, thanks for making CalendarDatePicker look good!
In overall this PR looks good to me as it is now, but I left couple of comments.

> The calendar button itself can be removed. This is now redundant functionality that isn't needed. In the UWP implementation it does not exist -- however, I currently keep it for backwards compatiblity.

You mean, specifically the button on the right of the control? Or the whole control?

> Keeping the calendar as a button may cause issues with navigation especially by keyboard.

I don't see any possible problems. Except it's not possible to "Shift+Tab" right now for some reason, is it related?

> There is an alternate way of constructing this control. Perhaps the entire control itself could just be a button. Then the TextBox is part of the content just like the icon. This might solve the above two issues. The style would have to be heavily customized though.

It feels like this control is primarily a text input control, and button is only an addition. Though if it would simplify control - sure.
To be honest right now this control looks just like a TextBox with a button set in InnerRightContent property. Like we have "clear" button there in default style. I wonder if it would make more sense to just use InnerRightContent with Button.Flyout.

robloo commented:
@maxkatz6 

> In overall this PR looks good to me as it is now, but I left couple of comments.

Thanks, I'll get to making the changes as soon as possible but it will be a little while.

> You mean, specifically the button on the right of the control? Or the whole control?

Yes, I'm talking about the button on the right of the control (now moved inside the border). The real issue is I don't know why a button is even needed. The UWP control template doesn't have a button. Everything is handled by events on the control itself. In fact I had to add some of that functionality so now we have duplicated code: when the user clicks on the CalendarDatePicker background and when the user clicks on the PART_Button are each separately handled. Also note WPF does have the button, not sure why it was done that way. 

Removing the PART_Button would simplify the template, code-behind and event handling. However, it would break compatibility with other styles that expect the button to be there. I'm just concerned it's too big of a breaking change for no functional benefit (only code clean-up).

> I don't see any possible problems. Except it's not possible to "Shift+Tab" right now for some reason, is it related?

For Focus/Navigation I just flagged it as a potential issue but didn't heavily test it.  Since the button isn't focusable I guess everything works well enough. I'll have to test more but generally I don't like having a full button like this in the template as it has the potential to mess things up. Other comments about this above/below.

>>There is an alternate way of constructing this control. Perhaps the entire control itself could just be a button. Then the TextBox is part of the content just like the icon. This might solve the above two issues. The style would have to be heavily customized though.
>
>It feels like this control is primarily a text input control, and button is only an addition. Though if it would simplify control - sure.

Yea, not a huge fan of this idea on second look myself.

> To be honest right now this control looks just like a TextBox with a button set in InnerRightContent property. Like we have "clear" button there in default style. I wonder if it would make more sense to just use InnerRightContent with Button.Flyout.

That's a very interesting idea to consider. I might play around with it but my feeling right now is just to remove the button completely. I'll keep thinking about this though, it may actually be the better solution to clean everything up in both the template and code-behind. (stop handling both the button + control itself separately) 

There is one more topic, should CalendarDatePicker be moved to it's own directory now -- either within Calendar or as a new control folder? I split several things out into separate files now which means this is probably a good idea to group them in a folder. There is already a lot within the Calendar folder itself.

robloo committed:
Simplify getting changed property values

Co-authored-by: Max Katz <maxkatz6@outlook.com>

robloo commented:
>> To be honest right now this control looks just like a TextBox with a button set in InnerRightContent property. Like we have "clear" button there in default style. I wonder if it would make more sense to just use InnerRightContent with Button.Flyout.
>
>That's a very interesting idea to consider. I might play around with it but my feeling right now is just to remove the button completely. I'll keep thinking about this though, it may actually be the better solution to clean everything up in both the template and code-behind. (stop handling both the button + control itself separately)

Thinking about this some more: It might be best to skip this type of change for now. It can be done in the future if it seems best. The fundamental concern I have now is while the CalendarDatePicker is currently editable with a TextBox as the fundamental control (like WPF) it may in the future have a non-editable state. The non-editable state may not have a TextBox at all (like in UWP). Therefore, basing the entire template around a TextBox as the fundamental control may make things more difficult in the future.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020375-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo commented on src/Avalonia.Controls/Calendar/CalendarDatePicker.cs:
```diff
@@ -538,16 +197,199 @@ protected override void OnApplyTemplate(TemplateAppliedEventArgs e)
+            // IsDropDownOpen
+            else if (change.Property == IsDropDownOpenProperty)
+            {
+                var oldValue = change.GetOldValue<bool>();
+                var value = change.GetNewValue<bool>();
```
Note for myself: Use `GetOldAndNewValue` like below.

robloo replied:
Updated

robloo commented on src/Avalonia.Controls/Calendar/CalendarDatePicker.cs:
```diff
@@ -538,16 +197,199 @@ protected override void OnApplyTemplate(TemplateAppliedEventArgs e)
+            // SelectedDate
+            else if (change.Property == SelectedDateProperty)
+            {
+                var addedDate = change.GetNewValue<DateTime?>();
+                var removedDate = change.GetOldValue<DateTime?>();
```
Note for myself: Use `GetOldAndNewValue` like below.

robloo replied:
Updated

robloo committed:
Simplify getting changed property values

robloo committed:
Merge branch 'master' into calendar-date-picker

robloo committed:
Remove width set in CalendarDatePickerPage

See: https://github.com/AvaloniaUI/Avalonia/issues/7502#issuecomment-1114673564

robloo commented:
I decided to keep PART_Button for now:
 1. It should be supported in code-behind anyway for the default and any custom app styles. Therefore, it wouldn't simplify any event handlers.
 2. It's actually one of the best containers for the calendar icon-related controls. That allows some nice simplifications in the styles to, for example, handle foreground color in the disabled state.
 3. It has the nice transform animations when pressed which is a better user experience.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020456-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Open Calendar popup with Alt+Down

robloo committed:
Simplify calendar popup toggling methods

robloo commented:
@maxkatz6 No further changes planned on my end. If it looks good to you, I'm ok to merge.

robloo committed:
Move CalendarDatePicker into its own directory

robloo commented:
Also note that the last commit moved this into its own `CalendarDatePicker` directory. If that is undesirable, it can be reverted. I leave the final decision up to you.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020459-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

maxkatz6 commented on src/Avalonia.Controls/CalendarDatePicker/CalendarDatePicker.cs:
```diff
@@ -576,78 +421,57 @@ protected override void OnGotFocus(GotFocusEventArgs e)
+            if ((key == Key.Space || key == Key.Enter) && IsEffectivelyEnabled) // Key.GamepadA is not currently supported
+            {
+                // Since the TextBox is used for direct date entry,
+                // it isn't supported to open the popup/flyout using these keys.
```
Just for case, it is possible to handle events with tunnel strategy (alternative of Preview events in uwp). 
Though I don't think it's really needed here for these keys.

robloo replied:
Yes, I think it's possible to get the actual events. I just don't think it's a good user experience. In this specific case, space is valid for a date and enter means 'commit and update' more than open a flyout. So I didnt enable this functionality and commented why. I think this applies to all input controls.

maxkatz6 committed:
Merge branch 'master' into calendar-date-picker

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020497-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

maxkatz6 commented:
I noticed only one problem with this PR, most likely unrelated to changes here.
Created a comment here https://github.com/AvaloniaUI/Avalonia/issues/8070#issuecomment-1131189197

robloo commented:
@maxkatz6 i noticed the same issue and was going to make the same note in text processing issues if it didn't dissapear soon. It is new and didn't exist last time I checked a few weeks ago. So we can probably narrow down the problem commit quickly.
